### DISCUSSION TITLE
Waitgroup in tokio!

### DISCUSSION THREAD

**Discussion by ab-basit** [tokio-rs/tokio#5319] [Ideas] [answered]
Can we have WaitGroup like feature in tokio?

**Comment by Darksonn** [comment, marked as answer]
The standard use-case for WaitGroup in go is to wait for a set of tasks to finish, but the best way to do that in Tokio is to store a `Vec<JoinHandle<T>>` and use a loop:
```Rust
let mut join_handles = Vec::new();
for i in 0..10 {
    join_handles.push(tokio::spawn(my_async_fn(i)));
}
for jh in join_handles {
    jh.await.unwrap();
}
```
You can also use a [`JoinSet`](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html) instead of the vector.

**Comment by ab-basit** [comment]
@Darksonn understood! but there could be complex workflow where we want to make sure all tasks are completed. 
I was thinking to have something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1707eedfa6f16ab7f30e49ee615f6d3c) which allows to clone the Taskgroup and pass around,  Finally wait on taskgroup , say in shutdown method where we want every task to complete before shutting down.

**Reply by Noah-Kennedy** [reply]
I feel like this calls for a new method in JoinSet.

**Comment by aaronriekenberg** [comment]
There is a library called awaitgroup that might be useful: https://crates.io/crates/awaitgroup

I am not the author, but I have used this in the past.  Has an interface similar to [`sync.WaitGroup`](https://pkg.go.dev/sync#WaitGroup) in go.

**Comment by laizy** [comment]
You can also try https://github.com/laizy/waitgroup-rs I made 3 years ago. It is tiny and performant : 
1. resuse the Arc's refcount to track the tasks instead another AtomicUsize;
2. lock free, do not use any Mutex in the implementation;
3. support nested tasks;

**Comment by Nugine** [comment]
A newer solution:
https://docs.rs/tokio-util/latest/tokio_util/task/task_tracker/struct.TaskTracker.html

**Comment by xrayw** [comment]
In addition to JoinSet and TaskTracker. we have another way to do it.

```rust
let (tx, rx) = mpsc::sync_channel::<i32>(1);
        
for i in 1..=5 {
    let tx1 = tx.clone();
    spawn(|| {
        // ...
        drop(tx1)
    });
}

drop(tx);

// it will block until all the senders disconnected, then return Err
let _ = rx.recv();

```
